source/MaterialXRuntime/RtFileIo.cpp

```diff
@@ -552,81 +572,63 @@ void RtFileIo::read(const FilePath& documentPath, const FileSearchPath& searchPa
         readOptions.skipConflictingElements = true;
         readFromXmlFile(document, documentPath, searchPaths, &readOptions);
 
-        CopyOptions copyOptions;
-        copyOptions.skipConflictingElements = true;
-
-        read(document, nullptr, filter);
+        PrvStage* stage = data()->asA<PrvStage>();
+        readDocument(document, stage, filter);
     }
     catch(Exception&)
     {
         return;
     }
 }
 
-void RtFileIo::loadLibraries(const StringVec& libraryPaths, const FileSearchPath& searchPaths)
+void RtFileIo::write(const FilePath& documentPath, const XmlWriteOptions* writeOptions, RtFileIo::WriteFilter filter)
 {
-    // We must create a single document with all dependents since the read
-    // functionality current looks for definitions within the same document.
-    DocumentPtr libraryDoc = createDocument();
-    StringVec libraryNames = MaterialX::loadLibraries(libraryPaths, searchPaths, libraryDoc);
-
     PrvStage* stage = data()->asA<PrvStage>();
-    for (const string& uri : libraryNames)
-    {
-        RtStage libraryStage = RtStage::createNew(RtToken(uri));
-        RtFileIo libStageIo(libraryStage.getObject());
-        RtStage searchStage(getObject());
-        libStageIo.read(libraryDoc, &searchStage,
-            [uri](const ElementPtr &e)
-        {
-            return (e->getActiveSourceUri() == uri);
-        });
-        stage->addReference(libStageIo.data());
-    }
+
+    DocumentPtr document = createDocument();
+    writeDocument(document, stage, filter);
+    writeToXmlFile(document, documentPath, writeOptions);
 }
 
-void RtFileIo::write(DocumentPtr& doc, RtFileIo::WriteFilter filter)
+void RtFileIo::loadLibraries(const StringVec& libraryPaths, const FileSearchPath& searchPaths)
 {
     PrvStage* stage = data()->asA<PrvStage>();
-    writeAttributes(stage, doc);
 
-    for (size_t i = 0; i < stage->numChildren(); ++i)
+    // Load all content into a document.
+    DocumentPtr doc = createDocument();
+    MaterialX::loadLibraries(libraryPaths, searchPaths, doc);
+
+    // First, load all nodedefs. Having these available is needed
+    // when node instances are loaded later.
+    for (const NodeDefPtr& nodedef : doc->getNodeDefs())
```

Comment:
Not sure if this was there before, but what happens to reading typedefs or is this just ignored?

Reply:
TypeDefs we are not able to read in data driven (we never have been), because for types we need more information that just the name. We need to register functions for parsing their values, and we need various syntax information registered on the Syntax classes. So new types can't be added in a purely data driven way unfortunately.